app/src/features/submissions/AdminSubmissionPage.tsx

```diff
@@ -1,115 +1,161 @@
+
+  const [isSecurityDialogOpen, setIsSecurityDialogOpen] = useState(false);
+
+  /* ---------------- Derived Data ---------------- */
+
+  const rows: FeatureRow[] = useMemo(() => {
+    return (
+      featureDataLoader.data?.features.map((feature) => ({
+        id: feature.submission_feature_id,
+        submission_feature_id: feature.submission_feature_id,
+        feature_type_display_name: feature.feature_type_name,
+        feature_type_name: feature.feature_type_name,
+        secured: feature.secured
+      })) ?? []
+    );
+  }, [featureDataLoader.data]);
+
+  const rowCount = featureDataLoader.data?.pagination.total ?? 0;
+
+  /* ---------------- Handlers ---------------- */
+
+  const toggleRowSelection = useCallback((params: GridRowParams<FeatureRow>) => {
+    const id = params.id as number;
 
-/**
- * AdminSubmissionPage component for reviewing submissions.
- *
- * @return {*}
- */
-const AdminSubmissionPage = () => {
-  const [groupedSubmissionFeatureSelection, setGroupedSubmissionFeatureSelection] =
-    useState<GroupedSubmissionFeatureSelection>({});
-
-  const submissionContext = useSubmissionContext();
-  const { submissionFeatureGroupsDataLoader } = submissionContext;
-  const submissionFeatureGroups = submissionFeatureGroupsDataLoader.data || [];
-
-  const onRowSelectionModelChange = (
-    featureTypeName: IGetSubmissionGroupedFeatureResponse['feature_type_name'],
-    rowSelectionModel: GridRowSelectionModel
-  ) => {
-    setGroupedSubmissionFeatureSelection((prev) => ({
-      ...prev,
-      [featureTypeName]: rowSelectionModel
-    }));
-  };
-
-  // Extract all selected feature IDs from the grouped selection
-  const submissionFeatureIds = Object.values(groupedSubmissionFeatureSelection).flatMap(extractIdsFromSelectionModel);
-
-  // Get all available feature IDs as fallback
-  const allSubmissionFeatureIds = submissionFeatureGroups.reduce(
-    (acc: number[], submissionFeatureGroup: IGetSubmissionGroupedFeatureResponse) => {
-      return acc.concat(submissionFeatureGroup.features.map((feature) => feature.submission_feature_id));
+    setSelectedFeatureIds((prev) => {
+      const next = new Set(prev);
+      next.has(id) ? next.delete(id) : next.add(id);
+      return next;
+    });
+  }, []);
+
+  const openSecurityDialog = useCallback((featureIds: Set<number>) => {
+    setDialogFeatureIds(featureIds);
+    setIsSecurityDialogOpen(true);
+  }, []);
+
+  const openSingleRowSecurity = useCallback(
+    (row: FeatureRow) => {
+      openSecurityDialog(new Set([row.submission_feature_id]));
     },
-    []
+    [openSecurityDialog]
   );
 
+  const openBulkSecurity = useCallback(() => {
+    openSecurityDialog(new Set(selectedFeatureIds));
+  }, [openSecurityDialog, selectedFeatureIds]);
+
+  const closeSecurityDialog = useCallback(() => {
+    setIsSecurityDialogOpen(false);
+  }, []);
+
+  const refreshFeatures = useCallback(() => {
```

Comment:
[SHOULD FIX] UI does not refresh after submission stage change. `refreshFeatures()` only refreshes `featureDataLoader` but not `submissionDataLoader`, so the header status badges don't update.

  ```suggestion
    const refreshFeatures = useCallback(() => {
      featureDataLoader.refresh(submissionId, featuresPagination);
      submissionDataLoader.refresh(submissionId);
    }, [featureDataLoader, submissionDataLoader, submissionId, featuresPagination]);